File: src/main/java/dev/xeonkryptos/xeonrobotframeworkplugin/config/RobotHighlighter.java
package dev.xeonkryptos.xeonrobotframeworkplugin.config;

import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.ExtendedRobotTypes;
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.RobotLexerAdapter;
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.RobotTypes;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class RobotHighlighter extends SyntaxHighlighterBase {

    public static final TextAttributesKey SECTION_TITLE = TextAttributesKey.createTextAttributesKey("ROBOT_SECTION_TITLE",
                                                                                                    DefaultLanguageHighlighterColors.STRING);
    public static final TextAttributesKey GLOBAL_SETTING_OPTION = TextAttributesKey.createTextAttributesKey("ROBOT_GLOBAL_SETTING_OPTION",
                                                                                                            DefaultLanguageHighlighterColors.DOC_COMMENT_TAG_VALUE);
    public static final TextAttributesKey LOCAL_SETTING_OPTION = TextAttributesKey.createTextAttributesKey("ROBOT_LOCAL_SETTING_OPTION",
                                                                                                           DefaultLanguageHighlighterColors.METADATA);
    public static final TextAttributesKey IMPORT = TextAttributesKey.createTextAttributesKey("ROBOT_IMPORT_OPTION",
                                                                                             DefaultLanguageHighlighterColors.DOC_COMMENT_TAG_VALUE);
    public static final TextAttributesKey IMPORT_ARGUMENT = TextAttributesKey.createTextAttributesKey("ROBOT_IMPORT_ARGUMENT",
                                                                                                      DefaultLanguageHighlighterColors.HIGHLIGHTED_REFERENCE);
    public static final TextAttributesKey USER_KEYWORD_NAME = TextAttributesKey.createTextAttributesKey("ROBOT_USER_KEYWORD_DEFINITION",
                                                                                                        DefaultLanguageHighlighterColors.KEYWORD);
    public static final TextAttributesKey TEST_CASE_NAME = TextAttributesKey.createTextAttributesKey("ROBOT_TEST_CASE_NAME",
                                                                                                     DefaultLanguageHighlighterColors.KEYWORD);
    public static final TextAttributesKey TASK_NAME = TextAttributesKey.createTextAttributesKey("ROBOT_TASK_NAME", DefaultLanguageHighlighterColors.KEYWORD);
    public static final TextAttributesKey KEYWORD = TextAttributesKey.createTextAttributesKey("ROBOT_KEYWORD_CALL",
                                                                                              DefaultLanguageHighlighterColors.FUNCTION_DECLARATION);
    public static final TextAttributesKey PARAMETER = TextAttributesKey.createTextAttributesKey("ROBOT_PARAMETER", DefaultLanguageHighlighterColors.PARAMETER);
    public static final TextAttributesKey ARGUMENT = TextAttributesKey.createTextAttributesKey("ROBOT_POSITIONAL_ARGUMENT",
                                                                                               DefaultLanguageHighlighterColors.STATIC_FIELD);
    public static final TextAttributesKey VARIABLE = TextAttributesKey.createTextAttributesKey("ROBOT_VARIABLE",
                                                                                               DefaultLanguageHighlighterColors.DOC_COMMENT_MARKUP);
    public static final TextAttributesKey EXTENDED_VARIABLE_ACCESS_BRACKETS = TextAttributesKey.createTextAttributesKey(
            "ROBOT_EXTENDED_VARIABLE_ACCESS_BRACKETS",
            DefaultLanguageHighlighterColors.BRACKETS);
    public static final TextAttributesKey COMMENT = TextAttributesKey.createTextAttributesKey("ROBOT_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT);
    public static final TextAttributesKey GHERKIN = TextAttributesKey.createTextAttributesKey("ROBOT_GHERKIN_KEYWORD",
                                                                                              DefaultLanguageHighlighterColors.METADATA);
    public static final TextAttributesKey STRUCTURAL_KEYWORDS = TextAttributesKey.createTextAttributesKey("ROBOT_STRUCTURAL_KEYWORD",
                                                                                                          DefaultLanguageHighlighterColors.METADATA);
    public static final TextAttributesKey PYTHON_EXPRESSION_CONTENT = TextAttributesKey.createTextAttributesKey("ROBOT_PYTHON_EXPRESSION_CONTENT",
                                                                                                                DefaultLanguageHighlighterColors.TEMPLATE_LANGUAGE_COLOR);
    public static final TextAttributesKey ERROR = TextAttributesKey.createTextAttributesKey(TokenType.BAD_CHARACTER.toString(),
                                                                                            DefaultLanguageHighlighterColors.INVALID_STRING_ESCAPE);

    public static final TextAttributesKey REASSIGNED_VARIABLE = TextAttributesKey.createTextAttributesKey("ROBOT_REASSIGNED_VARIABLE", VARIABLE);

    private static final Map<IElementType, TextAttributesKey> keys = new HashMap<>();

    static {
        keys.put(RobotTypes.SETTINGS_HEADER, SECTION_TITLE);
        keys.put(RobotTypes.VARIABLES_HEADER, SECTION_TITLE);
        keys.put(RobotTypes.USER_KEYWORDS_HEADER, SECTION_TITLE);
        keys.put(RobotTypes.COMMENTS_HEADER, SECTION_TITLE);
        keys.put(RobotTypes.TEST_CASES_HEADER, SECTION_TITLE);
        keys.put(RobotTypes.TASKS_HEADER, SECTION_TITLE);

        keys.put(RobotTypes.USER_KEYWORD_NAME, USER_KEYWORD_NAME);
        keys.put(RobotTypes.TEST_CASE_NAME, TEST_CASE_NAME);
        keys.put(RobotTypes.TASK_NAME, TASK_NAME);

        keys.put(RobotTypes.COMMENT, COMMENT);
        keys.put(RobotTypes.PARAMETER_NAME, PARAMETER);
        keys.put(RobotTypes.TEMPLATE_PARAMETER_NAME, PARAMETER);
        keys.put(RobotTypes.LITERAL_CONSTANT, ARGUMENT);
        keys.put(RobotTypes.TEMPLATE_ARGUMENT_VALUE, ARGUMENT);

        keys.put(RobotTypes.GIVEN, GHERKIN);
        keys.put(RobotTypes.WHEN, GHERKIN);
        keys.put(RobotTypes.THEN, GHERKIN);
        keys.put(RobotTypes.AND, GHERKIN);
        keys.put(RobotTypes.BUT, GHERKIN);

        keys.put(RobotTypes.FOR, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.FOR_IN, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.WHILE, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.TRY, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.EXCEPT, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.FINALLY, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.RETURN, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.GROUP, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.IF, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.ELSE_IF, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.ELSE, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.END, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.BREAK, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.CONTINUE, STRUCTURAL_KEYWORDS);
        keys.put(RobotTypes.WITH_NAME, STRUCTURAL_KEYWORDS);

        keys.put(RobotTypes.PYTHON_EXPRESSION_CONTENT, PYTHON_EXPRESSION_CONTENT);

        keys.put(RobotTypes.SCALAR_VARIABLE_START, VARIABLE);
        keys.put(RobotTypes.LIST_VARIABLE_START, VARIABLE);
        keys.put(RobotTypes.DICT_VARIABLE_START, VARIABLE);
        keys.put(RobotTypes.ENV_VARIABLE_START, VARIABLE);
        keys.put(RobotTypes.VARIABLE_BODY, VARIABLE);
        keys.put(RobotTypes.VARIABLE_END, VARIABLE);
        keys.put(ExtendedRobotTypes.EXTENDED_VARIABLE_ACCESS_BODY, VARIABLE);
        keys.put(RobotTypes.VARIABLE_ACCESS_START, EXTENDED_VARIABLE_ACCESS_BRACKETS);
        keys.put(RobotTypes.VARIABLE_ACCESS_END, EXTENDED_VARIABLE_ACCESS_BRACKETS);
        keys.put(RobotTypes.USER_KEYWORD_STATEMENT, USER_KEYWORD_NAME);
        keys.put(RobotTypes.KEYWORD_LIBRARY_NAME, KEYWORD);
        keys.put(RobotTypes.KEYWORD_LIBRARY_SEPARATOR, KEYWORD);
        keys.put(RobotTypes.KEYWORD_NAME, KEYWORD);

        keys.put(RobotTypes.SUITE_NAME_KEYWORD, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.DOCUMENTATION_KEYWORD, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.METADATA_KEYWORD, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.SETUP_TEARDOWN_STATEMENT_KEYWORDS, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.TAGS_KEYWORDS, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.TEMPLATE_KEYWORDS, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.TIMEOUT_KEYWORDS, GLOBAL_SETTING_OPTION);
        keys.put(RobotTypes.UNKNOWN_SETTING_KEYWORD, GLOBAL_SETTING_OPTION);

        keys.put(RobotTypes.LOCAL_SETTING_START, LOCAL_SETTING_OPTION);
        keys.put(RobotTypes.LOCAL_SETTING_END, LOCAL_SETTING_OPTION);
        keys.put(RobotTypes.LOCAL_SETTING_NAME, LOCAL_SETTING_OPTION);

        keys.put(RobotTypes.LIBRARY_IMPORT_KEYWORD, IMPORT);
        keys.put(RobotTypes.RESOURCE_IMPORT_KEYWORD, IMPORT);
        keys.put(RobotTypes.VARIABLES_IMPORT_KEYWORD, IMPORT);

        keys.put(TokenType.BAD_CHARACTER, ERROR);
    }

    @NotNull
    @Override
    public Lexer getHighlightingLexer() {
        return new RobotLexerAdapter();
    }

    @NotNull
    @Override
    public TextAttributesKey @NotNull [] getTokenHighlights(IElementType tokenType) {
        return pack(keys.get(tokenType));
    }
}

File: src/main/kotlin/dev/xeonkryptos/xeonrobotframeworkplugin/fileTypes/RobotHighlighter.kt
package dev.xeonkryptos.xeonrobotframeworkplugin.fileTypes

import com.intellij.lexer.Lexer
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.openapi.project.Project
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType
import dev.xeonkryptos.xeonrobotframeworkplugin.lexer.RobotLexerAdapter
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.ExtendedRobotTypes
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.RobotTypes

class RobotHighlighter @JvmOverloads constructor(private val project: Project? = null) : SyntaxHighlighterBase() {

    companion object {
        @JvmField
        val SECTION_TITLE: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_SECTION_TITLE", DefaultLanguageHighlighterColors.STRING)

        @JvmField
        val GLOBAL_SETTING_OPTION: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_GLOBAL_SETTING_OPTION", DefaultLanguageHighlighterColors.DOC_COMMENT_TAG_VALUE)

        @JvmField
        val LOCAL_SETTING_OPTION: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_LOCAL_SETTING_OPTION", DefaultLanguageHighlighterColors.METADATA)

        @JvmField
        val IMPORT_ARGUMENT: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_IMPORT_ARGUMENT", DefaultLanguageHighlighterColors.HIGHLIGHTED_REFERENCE)

        @JvmField
        val USER_KEYWORD_NAME: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_USER_KEYWORD_DEFINITION", DefaultLanguageHighlighterColors.KEYWORD)

        @JvmField
        val TEST_CASE_NAME: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_TEST_CASE_NAME", DefaultLanguageHighlighterColors.KEYWORD)

        @JvmField
        val TASK_NAME: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_TASK_NAME", DefaultLanguageHighlighterColors.KEYWORD)

        @JvmField
        val KEYWORD: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_KEYWORD_CALL", DefaultLanguageHighlighterColors.FUNCTION_DECLARATION)

        @JvmField
        val PARAMETER: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_PARAMETER", DefaultLanguageHighlighterColors.PARAMETER)

        @JvmField
        val ARGUMENT: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_POSITIONAL_ARGUMENT", DefaultLanguageHighlighterColors.STATIC_FIELD)

        @JvmField
        val VARIABLE: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_VARIABLE", DefaultLanguageHighlighterColors.DOC_COMMENT_MARKUP)
        val EXTENDED_VARIABLE_ACCESS_BRACKETS: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_EXTENDED_VARIABLE_ACCESS_BRACKETS", DefaultLanguageHighlighterColors.BRACKETS)

        @JvmField
        val COMMENT: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT)

        @JvmField
        val GHERKIN: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_GHERKIN_KEYWORD", DefaultLanguageHighlighterColors.KEYWORD)

        @JvmField
        val STRUCTURAL_KEYWORDS: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_STRUCTURAL_KEYWORD", DefaultLanguageHighlighterColors.KEYWORD)

        @JvmField
        val PYTHON_EXPRESSION_CONTENT: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_PYTHON_EXPRESSION_CONTENT", DefaultLanguageHighlighterColors.TEMPLATE_LANGUAGE_COLOR)
        val ERROR: TextAttributesKey = TextAttributesKey.createTextAttributesKey(TokenType.BAD_CHARACTER.toString(), DefaultLanguageHighlighterColors.INVALID_STRING_ESCAPE)

        @JvmField
        val REASSIGNED_VARIABLE: TextAttributesKey = TextAttributesKey.createTextAttributesKey("ROBOT_REASSIGNED_VARIABLE", VARIABLE)

        private val keys: MutableMap<IElementType?, TextAttributesKey?> = mutableMapOf()

        init {
            keys[RobotTypes.SETTINGS_HEADER] = SECTION_TITLE
            keys[RobotTypes.VARIABLES_HEADER] = SECTION_TITLE
            keys[RobotTypes.USER_KEYWORDS_HEADER] = SECTION_TITLE
            keys[RobotTypes.COMMENTS_HEADER] = SECTION_TITLE
            keys[RobotTypes.TEST_CASES_HEADER_NAME] = SECTION_TITLE
            keys[RobotTypes.TASKS_HEADER_NAME] = SECTION_TITLE
            keys[RobotTypes.TASKS_HEADER] = SECTION_TITLE

            keys[RobotTypes.DATA_DRIVEN_COLUMN_NAME] = SECTION_TITLE

            keys[RobotTypes.USER_KEYWORD_NAME_PART] = USER_KEYWORD_NAME
            keys[RobotTypes.TEST_CASE_NAME_PART] = TEST_CASE_NAME
            keys[RobotTypes.TASK_NAME_PART] = TASK_NAME

            keys[RobotTypes.COMMENT] = COMMENT
            keys[RobotTypes.PARAMETER_NAME] = PARAMETER
            keys[RobotTypes.TEMPLATE_PARAMETER_NAME] = PARAMETER
            keys[RobotTypes.LITERAL_CONSTANT] = ARGUMENT
            keys[RobotTypes.TEMPLATE_ARGUMENT_VALUE] = ARGUMENT

            keys[RobotTypes.GIVEN] = GHERKIN
            keys[RobotTypes.WHEN] = GHERKIN
            keys[RobotTypes.THEN] = GHERKIN
            keys[RobotTypes.AND] = GHERKIN
            keys[RobotTypes.BUT] = GHERKIN

            keys[RobotTypes.VAR] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.FOR] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.FOR_IN] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.FOR_IN_ENUMERATE] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.FOR_IN_RANGE] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.FOR_IN_ZIP] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.WHILE] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.TRY] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.EXCEPT] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.FINALLY] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.RETURN] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.GROUP] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.IF] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.ELSE_IF] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.ELSE] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.END] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.BREAK] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.CONTINUE] = STRUCTURAL_KEYWORDS
            keys[RobotTypes.WITH_NAME] = STRUCTURAL_KEYWORDS

            keys[RobotTypes.PYTHON_EXPRESSION_CONTENT] = PYTHON_EXPRESSION_CONTENT

            keys[RobotTypes.SCALAR_VARIABLE_START] = VARIABLE
            keys[RobotTypes.LIST_VARIABLE_START] = VARIABLE
            keys[RobotTypes.DICT_VARIABLE_START] = VARIABLE
            keys[RobotTypes.ENV_VARIABLE_START] = VARIABLE
            keys[RobotTypes.VARIABLE_BODY] = VARIABLE
            keys[RobotTypes.VARIABLE_LBRACE] = VARIABLE
            keys[RobotTypes.VARIABLE_RBRACE] = VARIABLE
            keys[RobotTypes.PYTHON_EXPRESSION_START] = VARIABLE
            keys[RobotTypes.PYTHON_EXPRESSION_END] = VARIABLE
            keys[ExtendedRobotTypes.EXTENDED_VARIABLE_ACCESS_BODY] = VARIABLE
            keys[RobotTypes.VARIABLE_ACCESS_START] = EXTENDED_VARIABLE_ACCESS_BRACKETS
            keys[RobotTypes.VARIABLE_ACCESS_END] = EXTENDED_VARIABLE_ACCESS_BRACKETS
            keys[RobotTypes.USER_KEYWORD_STATEMENT] = USER_KEYWORD_NAME
            keys[RobotTypes.KEYWORD_LIBRARY_NAME] = KEYWORD
            keys[RobotTypes.KEYWORD_LIBRARY_SEPARATOR] = KEYWORD
            keys[RobotTypes.KEYWORD_NAME] = KEYWORD

            keys[RobotTypes.SUITE_NAME_KEYWORD] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.DOCUMENTATION_KEYWORD] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.METADATA_KEYWORD] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.SETUP_TEARDOWN_STATEMENT_KEYWORDS] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.TAGS_KEYWORDS] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.TEMPLATE_KEYWORDS] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.TIMEOUT_KEYWORDS] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.UNKNOWN_SETTING_KEYWORD] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.LIBRARY_IMPORT_KEYWORD] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.RESOURCE_IMPORT_KEYWORD] = GLOBAL_SETTING_OPTION
            keys[RobotTypes.VARIABLES_IMPORT_KEYWORD] = GLOBAL_SETTING_OPTION

            keys[RobotTypes.LOCAL_SETTING_START] = LOCAL_SETTING_OPTION
            keys[RobotTypes.LOCAL_SETTING_END] = LOCAL_SETTING_OPTION
            keys[RobotTypes.LOCAL_SETTING_NAME] = LOCAL_SETTING_OPTION

            keys[TokenType.BAD_CHARACTER] = ERROR
        }
    }

    override fun getHighlightingLexer(): Lexer = RobotLexerAdapter(project)

    override fun getTokenHighlights(tokenType: IElementType?): Array<TextAttributesKey> = pack(keys[tokenType])
}
